### PR TITLE
Feature Intel Node Manager

### PR DESCRIPTION
**Pull request by trympet**
This pull request adds power metrics pulled from Intel Node Manager over IPMI.

#### New sensors:
image

I have placed the `IPMIClient` and `NodeManager` classes in the `LibreNMS\IPMI` namespace. Feel free to let me know if this is wrong. Also, I am not very familiar with PHP, so feel free to correct any misplaced naming conventions.

First time playing with the IPMI protocol; [the Intel Node Manager spec](https://www.intel.com/content/dam/www/public/us/en/documents/technical-specifications/intelligent-power-node-manager-specification.pdf) has proven itself quite useful. Seems there are very few tools which support Intel Node Manager despite it being quite an old spec. Regardless, this should provide some very useful data, allowing power metering on the component level  (CPU, RAM, Total (in watts)) on supported platforms.

The implementation is tested on the following supermicro hardware:
- X10DRW-I
- X8DTU-LN4F+
- X9SRL-F 
- X9SCL

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

#### Try it out
1. Add a new host
1. Under host configuration, click IPMI tab, enter IPMI host, username and password.
1. Run a discovery: `./discovery.php -h 5 -m sensors -d | grep -iF 'node manager'`. Output will tell you if it is supported or not.
1. Do a poll. Output should look something like this.
```
./poller.php -h 5 -m ipmi -d | grep -iF 'sensor intel'
Updating IPMI sensor Intel ME Memory... 9 Watts
Updating IPMI sensor Intel ME CPU... 35 Watts
Updating IPMI sensor Intel ME Platform...
```

#### Notes
- [x] General error handling
- [x] Detect NM version and end discovery / polling gracefully if not supported
- [x] Fetch data without enumerating entire SDR on every query
- [x] Connection timeouts 
- [ ] ~~Query IME for available sensors (OEM-specific; not reported by SDR)~~ defined in spec.
- [x] Add support for temp. sensors
- [x] Remove SDR cache dependency for distributed polling support
- [x] tests
- [x] Test case: non-supported SDR, v1 SDR, v2 SDR
- [x] Test case: malformed / non-conformant to spec. SDR 
- [x] Test case: empty SDR

The SDR should be stored in the database, given that the SDR is static. Future potential for performance improvements.

### CODE REVIEW

**Comment by trympet** [comment]
Todo:
- [ ] General error handling
- [ ] Detect NM version and end discovery / polling gracefully if not supported
- [ ] Fetch power data without enumerating entire SDR first if possible
- [ ] Connection timeouts 
- [ ] Query IME for available sensors (OEM-specific; not reported by SDR)
- [ ] Add support for temp. sensors

**Comment by Jellyfrog** [comment]
Hi

Initial look and it looks good 👍 
You should look into the errors reported by the tests below

**Comment by trympet** [comment]
@Jellyfrog, the 7.3 and linter tests should pass now

File: LibreNMS/IPMI/IPMIClient.php [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        $this->port = $port;
+    }
+
+    /**
+     * Gets a binary representation of the cached SDR record for this host.
+     * @return string|false The SDR binary or false on failure.
+     */
+    public function getSDR()
+    {
+        $path = '/tmp/librenms/ipmi/SDR';
```
Comment (Jellyfrog) [inline LibreNMS/IPMI/IPMIClient.php:101]: Should be in the cache dir in librenms root instead

File: LibreNMS/IPMI/IPMIClient.php [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        $this->assertSDR($path);
+
+        $filename = "$path/$this->host";
+        $sdrFile = fopen($filename, 'r');
+        try {
```
Comment (Jellyfrog) [inline LibreNMS/IPMI/IPMIClient.php:106]: You can replace all this with file_get_content

File: LibreNMS/IPMI/IPMIClient.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            if ($this->port) {
+                array_push($cmd, '-p', $this->port);
+            }
+
+            array_push($cmd, '-I', $this->interface);
+        }
+
+        $cmd = array_merge($cmd, explode(' ', $command));
+
+        return external_exec($cmd);
```
Comment (Jellyfrog) [inline LibreNMS/IPMI/IPMIClient.php:156]: Are you sure you want to use this function? It's very snmp centric; https://github.com/librenms/librenms/blob/5ba3df45eb5a351cec7f52a05c26d4936dcf4d0f/includes/common.php#L69

Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php:156]: Good point. I will create a slimmer version for ipmi execution. That way, I can also check and abort if the command errored.

File: LibreNMS/IPMI/IPMIClient.php (hunk 1 above)
Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Resolved in 88a8efd9834757c891497feb2de18f09aecd9a15.

File: LibreNMS/IPMI/IPMIClient.php (hunk 2 above)
Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Resolved in 88a8efd9834757c891497feb2de18f09aecd9a15.

File: LibreNMS/IPMI/IPMIClient.php (hunk 3 above)
Comment (Jellyfrog) [reply LibreNMS/IPMI/IPMIClient.php:159]: You can use https://github.com/librenms/librenms/blob/master/LibreNMS/Proc.php

Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Resolved in 03a28a6f81537fa7a1f44587bc20bb5c4baab9da.

**Comment by trympet** [comment]
> Please add test data so we can ensure your change is not broken in the future.
> 
> Read the docs to find out how: https://docs.librenms.org/Developing/os/Test-Units

The `IPMIClient` and `NodeManager` classes can be mocked, however, without dependency injection, it's hard to do any unit testing for the `ipmi.inc.php` poller and discovery files. 

Am I missing something? How is this typically handled? The first thing that comes to mind is to create a fake IPMI socket, the same way as snmpsim. Then perform assertions on the DB.

Meanwhile, I will create a fixture for the NodeManager class.

**Comment by Jellyfrog** [comment]
For me it looks good! Great work here. 
I think @murrant would be the best to review.

After this, if you would like to continue, you could convert impi module to the class based format. An example; https://github.com/librenms/librenms/pull/12605

**Comment by trympet** [comment]
Thank you! 😎 

Yes, refactoring the module to a class would make it testable. I will submit a new pr once I have time.

**Comment by trympet** [comment]
@murrant, review pretty please 😃

File: LibreNMS/IPMI/IPMICommand.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @link       https://www.librenms.org
+ * @copyright  2021 Trym Lund Flogard
+ * @author     Trym Lund Flogard <trym@flogard.no>
+ */
+
+namespace LibreNMS\IPMI;
+
+use Exception;
+use LibreNMS\Proc;
```
Comment (murrant) [inline LibreNMS/IPMI/IPMICommand.php:29]: Please use \Symfony\Component\Process\Process, LibreNMS\Proc is deprecated (should probably mark the class)

File: LibreNMS/IPMI/IPMIClient.php [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        $this->port = $port;
+    }
+
+    /**
+     * Gets a binary representation of the cached SDR record for this host.
+     * @return string|false The SDR binary or false on failure.
+     */
+    public function getSDR()
+    {
+        $basePath = Config::get('install_dir') . '/cache/ipmi';
```
Comment (murrant) [inline LibreNMS/IPMI/IPMIClient.php:104]: Maybe use the built in cache instead of a manual solution:
https://laravel.com/docs/cache

Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php:104]: Great idea! That should open it up for use with other ipmi functions as well, giving it a nice speed boost.

File: LibreNMS/IPMI/IPMIClient.php [hunk 6]
```diff
@@ -26,13 +26,15 @@
...
 use ErrorException;
-use LibreNMS\Config;
+use Illuminate\Support\Facades\Cache;
 
 /**
  * Represents an IPMI connection with a host machine.
  */
 class IPMIClient
 {
+    const SDR_TTL = 60 * 15;
```
Comment (trympet) [inline LibreNMS/IPMI/IPMIClient.php:36]: Should this be a config parameter?

File: LibreNMS/IPMI/IPMIClient.php (hunk 5 above)
Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Fixed in 3a144e5

File: LibreNMS/IPMI/IPMICommand.php (hunk 4 above)
Comment (trympet) [reply LibreNMS/IPMI/IPMICommand.php]: Fixed in 00abd18

File: LibreNMS/IPMI/IPMIClient.php (hunk 5 above)
Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Looks like the Laravel cache or the database messes with the binary data. Should probably b64 encode before storing it.

Comment (trympet) [reply LibreNMS/IPMI/IPMIClient.php]: Fixed 602ae29

File: LibreNMS/IPMI/IPMIClient.php (hunk 6 above)
Comment (murrant) [reply LibreNMS/IPMI/IPMIClient.php:36]: Try to minimize config variables in general.  They are easy to add later if needed.

**Comment by murrant** [comment]
It seems like SDR should be read during discovery and then polling should fetch based on SDR.
This would match the semantics of the rest of LibreNMS.

**Comment by murrant** [comment]
@trympet any plans to finish this?

**Comment by laf** [comment]
Sorry that we've not got this in a while @trympet. This PR now has a couple of conflicts since you created it, would you mind updating this PR and we can then take a proper look at it for merging?

**Comment by trympet** [comment]
> Sorry that we've not got this in a while @trympet. This PR now has a couple of conflicts since you created it, would you mind updating this PR and we can then take a proper look at it for merging?

Sure! I'd love some help to get this integrated!

File: includes/discovery/sensors/ipmi.inc.php [hunk 7]
```diff
@@ -64,6 +61,35 @@
...
     }
 
+    $nmClient = new NodeManager($client);
+    if ($nmClient->isPlatformSupported()) {
+        // Set Node Manager connection properties.
+        foreach ($nmClient->discoverAttributes() as $nmAttribKey => $nmAttribValue) {
+            set_dev_attrib($device, "node_manager_$nmAttribKey", $nmAttribValue);
+        }
+
+        $ipmi_unit_type = Config::get('ipmi_unit.Watts');
```
Comment (murrant) [inline includes/discovery/sensors/ipmi.inc.php:71]: ipmi_unit.Watts should be lowercase and added to config_definitions.  Same with other units, should be strtolower() to make them consistent.

File: LibreNMS/IPMI/IPMIClient.php [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    private function fetchSDR(): string
+    {
+        $basePath = sys_get_temp_dir() . '/ipmitool';
+        if (! is_dir($basePath)) {
+            mkdir($basePath, 0777, true);
+        }
+
+        $filePath = "$basePath/" . $this->host . '.sdr.tmp';
```
Comment (murrant) [inline LibreNMS/IPMI/IPMIClient.php:222]: $this->host needs to be escaped.

**Review by murrant — CHANGES_REQUESTED** [review]
Trying to merge some of these old PRs, I see a couple of issues and of course it needs a rebase again. :)

**Comment by murrant** [comment]
Closing due to lack of updates.  Feel free to re-open.

**Pull request closed without merging**